rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -58,40 +61,120 @@ public ObjectPoolPartition(ObjectPool<T> pool, PoolConfig config,
+      // Compensate for the removed object. Needed to prevent endless wait when in parallel a borrowObject is called
+      increaseObjects(1, false);
+      return;
+    }
+
+    log.debug(String.format("Returning object %s to queue of host %s. Queue size: %d", object, object.getHost(), objectQueue.size()));
+    try {
+      if (!objectQueue.offer(object, 5, TimeUnit.SECONDS)) {
+        log.error("Created more objects than configured: " + totalCount);
+        decreaseObject(object);
+      }
+    }
+    catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
   }
 
-  public synchronized int increaseObjects(int delta)
+  public Poolable<T> getObject(boolean blocking)
+  {
+    if (objectQueue.size() == 0) {
+      // increase objects and return one, it will return null if pool reaches max size or if object creation fails
+      Poolable<T> object = increaseObjects(this.config.getDelta(), true);
+
+      if (object != null) {
+        return object;
+      }
+
+      if (totalCount == 0) {
+        // Could not create objects, this is mostly due to connection timeouts hence no point blocking as there is not other producer of sockets
+        throw new RuntimeException("Could not add connections to pool");
+      }
+      // else wait for a connection to get free
+    }
+
+    Poolable<T> freeObject;
+    try {
+      if (blocking) {
+        freeObject = objectQueue.take();
+      }
+      else {
+        freeObject = objectQueue.poll(config.getMaxWaitMilliseconds(), TimeUnit.MILLISECONDS);
+        if (freeObject == null) {
+          throw new RuntimeException("Cannot get a free object from the pool");
+        }
+      }
+    }
+    catch (InterruptedException e) {
+      throw new RuntimeException(e); // will never happen
+    }
+
+    freeObject.setLastAccessTs(System.currentTimeMillis());
+    return freeObject;
+  }
+
+  private synchronized Poolable<T> increaseObjects(int delta, boolean returnObject)
   {
     int oldCount = totalCount;
     if (delta + totalCount > config.getMaxSize()) {
       delta = config.getMaxSize() - totalCount;
     }
+
+    Poolable<T> objectToReturn = null;
     try {
       for (int i = 0; i < delta; i++) {
         T object = objectFactory.create(host, socketTimeout, connectTimeout);
         if (object != null) {
-          objectQueue.put(new Poolable<>(object, pool, host));
+          // Do not put the first object on queue
+          // it will be returned to the caller to ensure it's request is satisfied first if object is requested
+          Poolable<T> poolable = new Poolable<>(object, pool, host);
+          if (objectToReturn == null && returnObject) {
+            objectToReturn = poolable;
+          }
+          else {
+            objectQueue.put(poolable);
+          }
           totalCount++;
         }
       }
       log.debug("Increased pool size by " + (totalCount - oldCount) + " to new size: " + totalCount + ", current queue size: " + objectQueue.size());
```

Comment:
should we add warn logs in case we are not able to increase the pool size?

Reply:
also, we can print delta as well, it may help us in debugging issues.

Reply:
adding

Reply:
done